### PR TITLE
 Sitemap filename can't exceed 32 characters #13937 

### PR DESCRIPTION
**Pull request by irajneeshgupta**
### Description (*)

The Sitemap filename length validation was not added , hence added `\Magento\Framework\Validator\StringLength` max 32 length validation for sitemap filename and it will not trim filename exceeding 32 chars. Client side validation in filename field of form is also added to fix this issue `'validate-length maximum-length-32'`. Updated `saveTest` unit test case.
### Fixed Issues (if relevant)

1. magento/magento2##13937: Issue title : Sitemap filename can't exceed 32 characters

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @irajneeshgupta. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by miguelbalparda** [comment]
Using the objectmanager doesn't seem to be the best path here, care to recheck?

**Comment by irajneeshgupta** [comment]
> Using the objectmanager doesn't seem to be the best path here, care to recheck?

@miguelbalparda
yeah but Magento itself used objectmanager the same in above lines for path validation , hence i added it in the same way.
I will recheck it and amend my commit.
Thanks for your suggestion.

**Comment by irajneeshgupta** [comment]
> Using the objectmanager doesn't seem to be the best path here, care to recheck?

@miguelbalparda 
I rechecked and removed objectManager from last commit. 
Thank you.

**Comment by irajneeshgupta** [comment]
@miguelbalparda can you please review changes.
Thanks.

**Review by aleron75 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @aleron75, thank you for the review.
[ENGCOM-3844](https://jira.corp.magento.com/browse/ENGCOM-3844) has been created to process this Pull Request

File: app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php [hunk 1]
```diff
@@ -6,11 +6,70 @@
...
 {
+    /**
+     * Maximum length of sitemap filename
+     */
+    const MAX_FILENAME_LENGTH = 32;
+
+    /**
+     * @var $_stringValidator
+     */
+    public $_stringValidator;
```
Comment (sivaschenko) [inline app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php]: Please change properties access level to private and remove the underscore prefix from property names

File: app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php [hunk 2]
```diff
@@ -6,11 +6,70 @@
...
+     * Save constructor.
+     * @param Action\Context $context
+     * @param \Magento\Framework\Validator\StringLength $stringValidator
+     * @param \Magento\MediaStorage\Model\File\Validator\AvailablePath $pathValidator
+     * @param \Magento\Sitemap\Helper\Data $sitemapHelper
+     * @param \Magento\Framework\Filesystem $filesystem
+     * @param \Magento\Sitemap\Model\SitemapFactory $sitemapFactory
+     */
+    public function __construct(
+        \Magento\Backend\App\Action\Context $context,
```
Comment (sivaschenko) [inline app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php]: Please add dependencies in a backward compatible way, see "Adding a constructor parameter" section of [Backward compatible development guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/#php) for details

File: app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php [hunk 3]
```diff
@@ -6,11 +6,70 @@
...
+ */
+class Save extends \Magento\Sitemap\Controller\Adminhtml\Sitemap implements HttpPostActionInterface
 {
+    /**
+     * Maximum length of sitemap filename
+     */
+    const MAX_FILENAME_LENGTH = 32;
+
+    /**
+     * @var $_stringValidator
```
Comment (sivaschenko) [inline app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php]: Please correct propeties' types in phpdocs

File: app/code/Magento/Sitemap/Test/Unit/Controller/Adminhtml/Sitemap/SaveTest.php [hunk 4]
```diff
@@ -50,8 +49,41 @@ class SaveTest extends \PHPUnit\Framework\TestCase
      */
     protected $messageManagerMock;
 
+    /**
+     * @var \Magento\Framework\Validator\StringLength|\PHPUnit_Framework_MockObject_MockObject
+     */
+    protected $lengthValidator;
```
Comment (sivaschenko) [inline app/code/Magento/Sitemap/Test/Unit/Controller/Adminhtml/Sitemap/SaveTest.php]: it's preferable to declare properties as private unless they are used by child classes

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for your contribution @irajneeshgupta , please see my code review notes.

File: app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php (hunk 1 above)
Comment (irajneeshgupta) [reply app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php]: Okay , I will made these changes.

File: app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php (hunk 2 above)
Comment (irajneeshgupta) [reply app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php]: okay I will update it.

File: app/code/Magento/Sitemap/Test/Unit/Controller/Adminhtml/Sitemap/SaveTest.php (hunk 4 above)
Comment (irajneeshgupta) [reply app/code/Magento/Sitemap/Test/Unit/Controller/Adminhtml/Sitemap/SaveTest.php]: okay sure

File: app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php (hunk 3 above)
Comment (irajneeshgupta) [reply app/code/Magento/Sitemap/Controller/Adminhtml/Sitemap/Save.php]: okay I will make these changes

**Comment by irajneeshgupta** [comment]
@sivaschenko 

I have made changes that you recommended.
Please review.

Thanks.

**Review by aleron75 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @aleron75, thank you for the review.
[ENGCOM-3844](https://jira.corp.magento.com/browse/ENGCOM-3844) has been created to process this Pull Request

**Comment by irajneeshgupta** [comment]
@sivaschenko 
any updates on this PR.

**Comment by sivaschenko** [comment]
Hi @irajneeshgupta thanks for your fixes, the pull request is being processed internally.

**Pull request merged**